docs/design/proposals/stateless-csi-plugin.md

```diff
@@ -0,0 +1,500 @@
+  across namespaces as in Alternative-1
+- Hence, move *ceph-cluster-secrets* to being mandatory in the StorageClass,
+  avoiding sharing cross-cluster secrets in a single secrets blob
+
+#### Pros alternative-2
+
+- Retains all of the prior pros in Alternative-1
+- Does not leak secrets across separate Ceph clusters, possibly managed by
+  different namespaces as there exists no *ceph-cluster-secrets*
+
+#### Cons alternative-2
+
+- *ceph-cluster-secrets* is absent and hence following operations **cannot**
+  be supported
+  - ListVolumes, ListSnapshots, and GetCapacity
+  - **NOTE:** Unless listing rbd and CephFS volumes can be done without
+    credentials
+- Bloats every StorageClass making secrets mandatory
+- Retains other cons as in Alternative-1
+
+### Solution Alternative-3
+
+#### Summary alternative-3
+
+- Address *ceph-clusters* being accessible to all namespaces that need to
+  manage it
+- Move the information in *ceph-clusters* to secrets in the StorageClass (as in
+  monValueFromSecret configuration implementation in the code as of this
+  writing)
+
+#### Pros and cons alternative-3
+
+- Cons remains the same, with further added complexity in the StorageClass and
+  its secrets
+- Removes the con in Alternate-1 and 2 where the *ceph-clusters* information is
+  accessible across namespaces
+
+### Solution Alternative-4
+
+#### Summary alternative-4
+
+- Create a CSI plugin instance per Ceph cluster, with its own unique provisioner
+  name
+- Each CSI plugin instance can now have its own *ceph-cluster* and
+  *ceph-secrets* config maps, where *ceph-cluster* contains information
+  regarding exactly one cluster and similarly the *ceph-secret* for the same
+- StorageClass points to the provisioner name and does not need secrets,
+  cluster-id and may optionally choose to contain pool-name (can be omitted,
+  if we choose to to allocate from an available pool in the cluster rather
+  than a specific pool, based on the CreateVolume request parameters and
+  Topology)
+
+#### Pros alternative-4
+
+- Can support all Operations, as cluster information and secrets are per
+  instance
+- The secrets and cluster information need not be shared across multiple
+  namespaces
+- StorageClass becomes leaner, but still needs at least as many StorageClasses
+  as there are clusters backing the CO environment
+- Allows for future merging of the CSI instances, reverting to the StorageClass
+  as in the proposed solution
+
+#### Cons alternative-4
+
+- We will have as many NodeService entities as there are clusters in each node
+  of the CO environment. This may consume more resources
+- We cannot, deal with multiple Ceph clusters backed by a single CSI plugin
+  instance, due to the strict 1:1 mapping between the same, and thus do any
+  form of intelligent provisioning across Ceph clusters for a request
+- **Needs Investigation** Does this hinder Topology based volume provisioning?
```

Comment:
Is a ceph cluster per topology how we would advocate deployment or would the ceph cluster span all topologies and use the crush map to direct allocations?

Reply:
I really do not have an answer to the question. I have made references to the same in @liewegas comment as well. Once I learn more about what Ceph can do, I can attempt to answer the question.

In my **assumption**, a Ceph cluster is per Topology.